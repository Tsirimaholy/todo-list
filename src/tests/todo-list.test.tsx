import React from "react";
import {create} from "react-test-renderer";
import TodoList, {TTask} from "../components/todo-list";
import {cleanup, fireEvent, render, screen} from "@testing-library/react";


afterEach(() => {
    cleanup();
})

describe('todo list component', function () {
    it('should render correctly', function () {
        const renderer = create(<TodoList taskList={[]}/>);
        expect(renderer.toJSON()).toMatchSnapshot();
    });

    it('should contain all provided tasks', function () {
        const TODO_LIST_MOCK: TTask[] = [
            {name: "todo1", done: false},
            {name: "todo2", done: true},
            {name: "todo3", done: true}
        ]

        const EXPECTED_NUMBER_OF_TASKS = 3;

        render(<TodoList taskList={TODO_LIST_MOCK}/>)
        const {children: {length: actualNumbersOfTasks}} = screen.getByRole("list") as HTMLUListElement;
        expect(actualNumbersOfTasks).toEqual(EXPECTED_NUMBER_OF_TASKS);
    });


    it('should call the onInsert func when the "Enter" key is pressed', function () {
        const TODO_LIST_MOCK: TTask[] = [
            {name: "todo1", done: false},
            {name: "todo2", done: true},
            {name: "todo3", done: true}
        ]
        const onInsertMock = jest.fn();

        render(<TodoList taskList={TODO_LIST_MOCK} onInsert={onInsertMock}/>);

        const textInput = screen.getByRole("textbox");

        fireEvent.keyDown(textInput, {key: "Enter"});
        expect(onInsertMock).toHaveBeenCalled();
    });

    it('should call the onDone func when the checkbox is clicked', function () {
        const TODO_LIST_MOCK: TTask[] = [
            {name: "todo1", done: false},
        ]
        const onDoneMock = jest.fn();

        render(<TodoList taskList={TODO_LIST_MOCK} onDone={onDoneMock}/>);


        const firstTodoCheckBox = screen.getAllByRole("checkbox")[0];
        fireEvent.click(firstTodoCheckBox);

        expect(onDoneMock).toHaveBeenCalled();
    });
});