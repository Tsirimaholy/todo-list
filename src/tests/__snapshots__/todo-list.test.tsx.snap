// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo list component should render correctly 1`] = `
<div
  className="task_list__container"
>
  <div
    className="todo_list__form"
  >
    <input
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="todo-input"
      type="text"
      value=""
    />
  </div>
  <ul
    className="task_list"
  />
</div>
`;
